File: 91-Decode Ways.py
class Solution(object):
    def numDecodings(self, s):
        """
        :type s: str
        :rtype: int
        """
        n = len(s)
        dp = [0]*(n+1)
        dp[0] = 1
        
        for i in range(len(s)):
            if 1 <= int(s[i]) <= 9:
                dp[i+1] += dp[i]
            if i > 0:
                if 10 <= int(s[i-1:i+1]) <= 26:
                    dp[i+1] += dp[i-1]
            if dp[i+1] == 0:
                return 0

        return dp[-1]

File: C++/91-Decode Ways.cpp
class Solution {
public:
    int numDecodings(string s) {
        int N = s.size();
        vector<int> dp(N+1, 0);
        dp[0] = 1;
        for (int i=0; i<N; i++){
            if (i > 0 && 10 <= stoi(s.substr(i-1, 2)) && stoi(s.substr(i-1, 2)) <= 26){
                dp[i+1] += dp[i-1];
            }
            if (s[i] <= '9' && s[i] > '0'){
                dp[i+1] += dp[i];
            }
        }
        return dp[dp.size()-1];
    }
};